### PR TITLE
Data Pipeline Standardization

### LINKED ISSUE #1538: Feature: Tweaks to Core Analysis Pipeline
## Feature Details

Update the v2 data pipeline to standardize naming and drop extra columns. Updated spec is as follows: 

Data Pipeline Specifications

The following section provides a detailed specifications for the required columns for the data pipeline used in the `{gsm}` package.The following tables are used in the data pipeline:

-   `dfInput`: Input data; Cross-domain participant-level input data with all needed data for KRI derivation.
-   `dfTransformed`: Transformed data; Group-level transformed data including KRI calculation.
-   `dfAnalyzed`: Analyzed data; Group-level analysis results data with `Score` column added providing a statistical score for the group.
-   `dfFlagged`: Flagged data; Group-level analysis results with `Flag` column added to indicate potential statistical outliers.
-  `dfSummary` : Summary data. Can add needed columns here. 

Many columns are shared across tables

| Table    | Column Name  | Description                          | Type     | Optional  | 
|----------|--------------|--------------------------------------|----------|--|
| All      | GroupID      | The group ID for the metric          | Character| | |
| All      | GroupType      | The group type for the metric (e.g. "Site")    | Character| | 
| All      | Numerator    | The calculated numerator value       | Numeric  | |
| All      | Denominator  | The calculated denominator value     | Numeric | |
| All      | Metric       | The calculated rate/metric value     | Numeric  | |
| dfInput  | SubjectID    | The subject ID                       | Character| |
| dfAnalyzed/dfFlagged | Score        | The Stastical Score      | Numeric  | |
| dfFlagged | Flag         | The Flag value                      | Numeric  | |
| dfSummary      | MetricID     | The Metric ID                        | Character| * |
| dfSummary     | StudyID      | The Study ID                         | Character| * |
| dfSummary      | SnapshotDate | The Date of the snapshot             | Date     | * |
 
Note that additional columns are ok in most data sets, but the columns listed are required unless noted. 

## Possible Implementation

Changes for implementation include: 
- Replace SiteID/CountryID with GroupID throughout.
- Add `GroupType` throughout.
- Replace "Rate" with the more general "Metric" throughout. 
- Add process to add additional columns to dfSummary

### PR DESCRIPTION
**Pull request by jwildfire**
## Overview

Workflow refactor per #1538. 

Additional tweaks include: 
- moved `RunQuery` calls from metric workflows to a new `mapping.yaml` that should be run before the metrics. 
- added  `strMetricID` to `Visualize_Metric` to simplify calling for multiple KRIs in `map()`.
- refactored parameter handling in `RunStep`.
- improved console messages in `RunStep` and `RunWorkflow`.

closes #1538.

## Test Notes/Sample Code

Test code for KRI Workflows 

```
# 1 - AE KRI Workflow
dfInput <- Input_Rate(
    dfSubjects= clindata::rawplus_dm,
    dfNumerator= clindata::rawplus_ae,
    dfDenominator = clindata::rawplus_dm,
    strSubjectCol = "subjid",
    strGroupCol = "siteid",
    strNumeratorMethod= "Count",
    strDenominatorMethod= "Sum",
    strDenominatorCol= "timeonstudy"
)  

dfTransformed <- Transform_Rate(dfInput)  
dfAnalyzed <- Analyze_NormalApprox(dfTransformed, strType = "rate") 
dfFlagged <- Flag_NormalApprox(dfAnalyzed, vThreshold = c(-3,-2,2,3))
dfSummarized <- Summarize(dfFlagged)
table(dfSummarized$Flag)

# 2 - Generate an SAE KRI from raw data using pipes
SAE_KRI <- Input_Rate(
    dfSubjects= clindata::rawplus_dm,
    dfNumerator= clindata::rawplus_ae %>% filter(aeser=="Y"),
    dfDenominator = clindata::rawplus_dm,
    strSubjectCol = "subjid",
    strGroupCol = "siteid",
    strNumeratorMethod= "Count",
    strDenominatorMethod= "Sum",
    strDenominatorCol= "timeonstudy"
) %>% 
Transform_Rate %>% 
Analyze_NormalApprox(strType = "rate") %>%
Flag_NormalApprox(vThreshold = c(-3,-2,2,3)) %>%
Summarize

table(SAE_KRI$Flag)
```

Test code for KRI Report

```
devtools::load_all()

# Prepare Workflows
wf_mapping <- MakeWorkflowList(strNames="mapping")[[1]]
wf_metrics <- MakeWorkflowList(strNames=paste0("kri",sprintf("%04d", 1:12)))
dfMetrics <- wf_metrics %>% map_df(~.x$meta)

# Import Site+Study Metadata
dfStudy<-clindata::ctms_study %>% rename(StudyID = protocol_number)
dfSite<- clindata::ctms_site %>% rename(SiteID = site_num)

# Pull Raw Data - this will overwrite the previous data pull
lRaw <- gsm::UseClindata(
  list(
    "dfSUBJ" = "clindata::rawplus_dm",
    "dfAE" = "clindata::rawplus_ae",
    "dfPD" = "clindata::ctms_protdev",
    "dfLB" = "clindata::rawplus_lb",
    "dfSTUDCOMP" = "clindata::rawplus_studcomp",
    "dfSDRGCOMP" = "clindata::rawplus_sdrgcomp",
    "dfDATACHG" = "clindata::edc_data_points",
    "dfDATAENT" = "clindata::edc_data_pages",
    "dfQUERY" = "clindata::edc_queries",
    "dfENROLL" = "clindata::rawplus_enroll"
  )
)

# Create Mapped Data
lMapped <- RunWorkflow(lWorkflow = wf_mapping, lData = lRaw)$lData

# Run Metrics
lResults <-wf_metrics %>% map(~RunWorkflow(lWorkflow=.x, lData=lMapped))

dfBounds <- lResults %>% 
  imap_dfr(~.x$lData$dfBounds %>% mutate(MetricID = .y)) %>%
  mutate(StudyID = "ABC-123") %>%
  mutate(SnapshotDate = Sys.Date())

dfSummary <- lResults %>% 
  imap_dfr(~.x$lData$dfSummary %>% mutate(MetricID = .y)) %>%
  mutate(StudyID = "ABC-123") %>%
  mutate(SnapshotDate = Sys.Date())

# Create charts
lCharts <- unique(dfSummary$MetricID) %>% map(~Visualize_Metric(
    dfSummary = dfSummary, 
    dfBounds = dfBounds, 
    dfSite = dfSite,
    dfMetrics = dfMetrics, 
    strMetricID = .x
  )
) %>% setNames(unique(dfSummary$MetricID)) 

# Run reports 
strOutpath <- "/Users/jwildfire/Documents/github/gsm-playground/gsm_site_report_db.html"
Report_KRI( lCharts = lCharts, dfSummary = dfSummary,  dfSite = dfSite, dfStudy = dfStudy, dfMetrics =dfMetrics, strOutpath = strOutpath )
```

### CODE REVIEW

**Comment by jwildfire** [comment]
@lauramaxwell - want to take a shot at updating the country workflows in a new commit to this branch?

**Comment by lauramaxwell** [comment]
> @lauramaxwell - want to take a shot at updating the country workflows in a new commit to this branch?

sure!

File: inst/report/Report_KRI.Rmd [hunk 1]
```diff
@@ -86,7 +86,7 @@ Report_OverviewText(
 for (i in seq_along(lCharts)) {
   MetricID <- names(lCharts)[[i]]
   if(!is.null(dfMetrics)){
-    MetricName <- dfMetrics$metric[dfMetrics$workflowid == MetricID]
+    MetricName <- dfMetrics$metric[dfMetrics$MetricID == MetricID]
```
Comment (lauramaxwell) [inline inst/report/Report_KRI.Rmd:89]: i'm getting an error: `Error in writeImpl(): ! Text to be written must be a length-one character vector`, so i think we might need to add a `%>% unique()` to the end of this to ensure MetricName is length one.

**Review by lauramaxwell — COMMENTED** [review]
one small thing i found in the Report_KRI() call (see comment in code). otherwise looking good!

**Comment by lauramaxwell** [comment]
@jwildfire @samussiah - made a few small edits to the country workflows to get them working. here is a script to test them out. basically the same as the kri script above with a few small tweaks for country-level reporting.
```
wf_mapping <- MakeWorkflowList(strNames="mapping")[[1]]
wf_metrics <- MakeWorkflowList(strNames=paste0("cou",sprintf("%04d", 1:12)))
dfMetrics <- wf_metrics %>% map_df(~.x$meta)

# Import Site+Study Metadata
dfStudy<-clindata::ctms_study %>% rename(StudyID = protocol_number)
dfSite<- clindata::ctms_site %>% rename(SiteID = site_num)

# Pull Raw Data - this will overwrite the previous data pull
lRaw <- gsm::UseClindata(
  list(
    "dfSUBJ" = "clindata::rawplus_dm",
    "dfAE" = "clindata::rawplus_ae",
    "dfPD" = "clindata::ctms_protdev",
    "dfLB" = "clindata::rawplus_lb",
    "dfSTUDCOMP" = "clindata::rawplus_studcomp",
    "dfSDRGCOMP" = "clindata::rawplus_sdrgcomp",
    "dfDATACHG" = "clindata::edc_data_points",
    "dfDATAENT" = "clindata::edc_data_pages",
    "dfQUERY" = "clindata::edc_queries",
    "dfENROLL" = "clindata::rawplus_enroll"
  )
)

# Create Mapped Data
lMapped <- RunWorkflow(lWorkflow = wf_mapping, lData = lRaw)$lData

# Run Metrics
lResults <-wf_metrics %>% map(~RunWorkflow(lWorkflow=.x, lData=lMapped))

dfBounds <- lResults %>% 
  imap_dfr(~.x$lData$dfBounds %>% mutate(MetricID = .y)) %>%
  mutate(StudyID = "ABC-123") %>%
  mutate(SnapshotDate = Sys.Date())

dfSummary <- lResults %>% 
  imap_dfr(~.x$lData$dfSummary %>% mutate(MetricID = .y)) %>%
  mutate(StudyID = "ABC-123") %>%
  mutate(SnapshotDate = Sys.Date())


# Create charts
lCharts <- unique(dfSummary$MetricID) %>% map(~Visualize_Metric(
  dfSummary = dfSummary, 
  dfBounds = dfBounds, 
  dfSite = dfSite,
  dfMetrics = dfMetrics, 
  strMetricID = .x
)
) %>% setNames(unique(dfSummary$MetricID)) 

# Run reports 
strOutpath <- "~/gsm_country_report_db.html"
Report_KRI( lCharts = lCharts, dfSummary = dfSummary,  dfSite = dfSite, dfStudy = dfStudy, dfMetrics =dfMetrics, strOutpath = strOutpath )
```

**Review by samussiah — APPROVED** [review]

File: inst/report/Report_KRI.Rmd (hunk 1 above)
Comment (jwildfire) [reply inst/report/Report_KRI.Rmd]: just need to use this for the dfMetrics derivation now: 

```
dfMetrics <- wf_metrics %>% map_df(function(wf){
  wf$meta$vThreshold <- paste(wf$meta$vThreshold, collapse = ",")
  return(wf$meta)
})
```

**Pull request merged**